src/core/comm/coll.cc

```diff
@@ -0,0 +1,389 @@
+  *id = current_unique_id;
+  current_unique_id++;
+  assert(current_unique_id <= MAX_NB_COMMS);
+  return CollSuccess;
+}
+
+#ifdef LEGATE_USE_GASNET
+MPI_Datatype dtypeToMPIDtype(CollDataType dtype)
+{
+  MPI_Datatype mpi_dtype = MPI_BYTE;
+  switch (dtype) {
+    case CollDataType::CollInt8: mpi_dtype = MPI_INT8_T; break;
+    case CollDataType::CollChar: mpi_dtype = MPI_CHAR; break;
+    case CollDataType::CollUint8: mpi_dtype = MPI_UINT8_T; break;
+    case CollDataType::CollInt: mpi_dtype = MPI_INT; break;
+    case CollDataType::CollUint32: mpi_dtype = MPI_UINT32_T; break;
+    case CollDataType::CollInt64: mpi_dtype = MPI_INT64_T; break;
+    case CollDataType::CollUint64: mpi_dtype = MPI_UINT64_T; break;
+    case CollDataType::CollFloat: mpi_dtype = MPI_FLOAT; break;
+    case CollDataType::CollDouble: mpi_dtype = MPI_DOUBLE; break;
+    default: log_coll.fatal("Unknown datatype"); assert(0);
+  }
+  return mpi_dtype;
+}
+
+static inline int match2ranks(int rank1, int rank2)
+{
+  // tag: seg idx + rank_idx + tag
+  // send_tag = sendto_global_rank * 10000 + global_rank (concat 2 ranks)
+  // which dst seg it sends to (in dst rank)
+  // recv_tag = global_rank * 10000 + recvfrom_global_rank (concat 2 ranks)
+  // idx of current seg we are receving (in src/my rank)
+  // example:
+  // 00 | 01 | 02 | 03
+  // 10 | 11 | 12 | 13
+  // 20 | 21 | 22 | 23
+  // 30 | 31 | 32 | 33
+  // 01's send_tag = 10, 10's recv_tag = 10, match
+  // 12's send_tag = 21, 21's recv_tag = 21, match
+
+  int tag;
+  constexpr int const max_ranks = 10000;
+  tag                           = rank1 * max_ranks + rank2;
+
+  // still under testing, will be used once if runs out of tags
+  // constexpr int const max_threads = 128;
+  // tag = rank1 % max_threads * max_ranks + rank2;
+
+  // if (rank1 >= rank2) {
+  //   tag = rank1*rank1 + rank1 + rank2;
+  // } else {
+  //   tag = rank1 + rank2*rank2;
+  // }
+  // tag = (rank1 + rank2) * (rank1 + rank2 + 1) / 2 + rank1;
+
+  return tag;
+}
+
+int generateAlltoallTag(int rank1, int rank2, CollComm global_comm)
+{
+  int tag = match2ranks(rank1, rank2) * CollTag::MAX_TAG + CollTag::ALLTOALL_TAG;
+  assert(tag < INT_MAX && tag > 0);
+  return tag;
+}
+
+int generateAlltoallvTag(int rank1, int rank2, CollComm global_comm)
+{
+  int tag = match2ranks(rank1, rank2) * CollTag::MAX_TAG + CollTag::ALLTOALLV_TAG;
+  assert(tag < INT_MAX && tag > 0);
+  return tag;
+}
+
+int generateBcastTag(int rank, CollComm global_comm)
+{
+  int tag = rank * CollTag::MAX_TAG + CollTag::BCAST_TAG;
+  assert(tag < INT_MAX && tag >= 0);
+  return tag;
+}
+
+int generateGatherTag(int rank, CollComm global_comm)
+{
+  int tag = rank * CollTag::MAX_TAG + CollTag::GATHER_TAG;
+  assert(tag < INT_MAX && tag > 0);
+  return tag;
+}
+
+#else  // undef LEGATE_USE_GASNET
+size_t getDtypeSize(CollDataType dtype)
+{
+  size_t size = 0;
+  switch (dtype) {
+    case CollDataType::CollInt8:
+    case CollDataType::CollChar: size = sizeof(char); break;
```

Comment:
You can directly return the value, instead of going through this assignment; i.e.,
```
    case CollDataType::CollChar: return sizeof(char);
```
Also, please use braces for the body of each case.

Reply:
fixed